rts/lra/test/arquillian-extension/src/main/java/io/narayana/lra/arquillian/spi/NarayanaLRARecovery.java

```diff
@@ -60,48 +64,86 @@ public void waitForCallbacks(URI lraId) {
      *
-     * @param host  the LRA coordinator host address
-     * @param port  the LRA coordinator port
      * @param lraId the LRA id of the LRA that is intended to be recovered
      * @return true the intended LRA recovered, false otherwise
      */
-    private boolean recoverLRAs(String host, int port, URI lraId) {
+    private boolean recoverLRAs(URI lraId) {
         // trigger a recovery scan
         Client recoveryCoordinatorClient = ClientBuilder.newClient();
 
         try {
-            String recoveryCoordinatorUrl = String.format("http://%s:%d/%s/recovery",
-                host, port, LRAConstants.RECOVERY_COORDINATOR_PATH_NAME);
-            WebTarget recoveryTarget = recoveryCoordinatorClient.target(URI.create(recoveryCoordinatorUrl));
+            URI lraCoordinatorUri = extractLRACoordinatorUri(lraId);
+            URI recoveryCoordinatorUri = UriBuilder.fromUri(lraCoordinatorUri)
+                    .path(RECOVERY_COORDINATOR_PATH_NAME).build();
+            WebTarget recoveryTarget = recoveryCoordinatorClient.target(recoveryCoordinatorUri);
 
             // send the request to the recovery coordinator
             Response response = recoveryTarget.request().get();
             String json = response.readEntity(String.class);
             response.close();
 
-            if (json.contains(lraId.toASCIIString())) {
-                // intended LRA didn't recover
-                return false;
-            }
-
-            return true;
+            // intended LRA didn't recover
+            return !json.contains(lraId.toASCIIString());
         } finally {
             recoveryCoordinatorClient.close();
         }
+    }
 
+    /**
+     * <p>
+     * This method extracts the coordinator URI from the provided LRA id.
+     * The Narayana LRA id consists of an URI of the LRA coordinator followed with the LRA transaction identifier.
+     * We know the LRA coordinator API is hardcoded with path {@value io.narayana.lra.LRAConstants#COORDINATOR_PATH_NAME}
+     * and the string will be searched in the original LRA id for.
+     * </p>
+     * <p>
+     * The Narayana LRA works with HTTP requests and URL is used for identification. An example of the standard pattern
+     * of the provided LRA id URI is:
+     * {@code http://localhost:8080/deployment/lra-coordinator/0_ffff0a28054b_9133_5f855916_a7}.
+     * This told us that the LRA can be accessed at the provided address by HTTP call.
+     * The LRA coordinator API is available at {@code http://localhost:8080/deployment/lra-coordinator}
+     * and the {@code 0_ffff0a28054b_9133_5f855916_a7} is the LRA transaction identifier used inside of the Coordinator
+     * to differentiate the LRA instances.
+     * </p>
+     *
+     * @param lraId  LRA id URI which is used for extraction of the LRA Coordinator API address
+     * @return LRA Coordinator API address
+     */
+    static URI extractLRACoordinatorUri(URI lraId) {
```

Comment:
... and the implementation is even harder to understand.

Reply:
I cleaned it as well.

Reply:
And it is 35 lines long. That seems excessive since there is a simple one liner (or some variant of it):
    `String getUid(URI lraId) {return lraId.getPath().substring(path.lastIndexOf('/') + 1);`

Reply:
I agree with Mike here, we could use the more efficient method and could give the explanation in java docs?

Reply:
Ok. I changed the method. It's not a one liner but it's shorter and similar.